R/extract.R

```diff
@@ -31,28 +38,69 @@ st_extract = function(x, ...) UseMethod("st_extract")
 #' st_extract(r, pnt) %>% st_as_sf()
 #' st_extract(r[,,,1], pnt)
 #' st_extract(r, st_coordinates(pnt)) # "at" is a matrix: return a matrix
+#' # Extraction on non-POINT geometries
+#' poly = st_buffer(pnt, 1000)
+#' st_extract(r, poly)
+#' 
+#' # Extraction with time matching
+#' rdate = c(r, r*2, along = "date")
+#' dates = c(Sys.Date()-1, Sys.Date())
+#' rdate = st_set_dimensions(rdate, "date", values = c(dates))
+#' 
+#' pntsf = st_sf(date = dates, geometry = pnt)
+#' st_extract(split(rdate, "band"), pntsf, time_column = "date") # POINT geometries
+#' 
+#' polysf = st_buffer(pntsf, 1000)
+#' st_extract(split(rdate, "band"), polysf, time_column = "date") # POLYGON geometries
+#' 
+#' vdc = st_sf(rdm = rnorm(20), polygons = st_buffer(st_sample(st_bbox(pnt), 20), 500),
+#' 			geometry = rep(pnt, 2), date = rep(dates, each = 10)) |> 
+#' 	st_as_stars(dims = c("geometry", "date"))
+#' 
+#' (vdc_new = st_extract(split(rdate, "band"), vdc, time_column = "date")) # stars vector data cube
+#' merge(vdc_new, name = "band")
+#' 
+#' ### Extraction applied to the geometries inside the vector data cube (cell values)
+#' (vdc_new2 = st_extract(split(rdate, "band"), vdc, time_column = "date", 
+#' 					   sfc_attribute = "polygons")) # stars vector data cube
+#' merge(vdc_new2, name = "band")
 st_extract.stars = function(x, at, ..., bilinear = FALSE, time_column = 
 		attr(at, "time_column") %||% attr(at, "time_col"),
 		interpolate_time = bilinear, FUN = mean,
-		resampling = c("nearest", "bilinear", "cubic", "cubicspline")) {
+		resampling = c("nearest", "bilinear", "cubic", "cubicspline"),
+		sfc_attribute = NULL) {
 
-	stopifnot(inherits(at, c("sf", "sfc", "matrix")))
+	stopifnot(inherits(at, c("sf", "sfc", "stars", "matrix")))
 	resampling = match.arg(resampling)
 	if (bilinear) {
 		stopifnot(resampling %in% c("nearest", "bilinear"))
 		resampling = "bilinear"
 	}
+	at_orig = at
+	if (inherits(at_orig, "stars") & !is.null(time_column)) {
+		at = st_as_sf(at, long = TRUE) 
+		if (!is.null(sfc_attribute)) {
+			sfc_dim = st_geometry(at)
+			sfc_dim_name = attr(at, "sf_column")
+			if(is.null(at[[sfc_attribute]])) at else st_geometry(at) = sfc_attribute
```

Comment:
It is not clear what returning `at` in the `if` TRUE branch does: `at` is not assigned somewhere?

Reply:
`at` remains the same, what happens in line 85 is the assignment of the active geometry. If the `sfc_attribute` is not a column found on the resulting sf object, then it will not reassign the geometry, if it is, it activates the geometry in the `sfc_attribute` column